File: 2300/main.go
package main

import (
	"leetcode/utils"
	"sort"
)

// 28 ms / 14.56 MB
func successfulPairs(spells []int, potions []int, success int64) []int {
	sort.Ints(potions)
	num_success := make([]int, len(spells))

	for i, spell := range spells {
		num_success[i] = search(potions, spell, success)
	}

	return num_success
}

func search(potions []int, spell int, success int64) int {
	left, mid := 0, 0
	right := len(potions) - 1

	for left <= right {
		mid = left + (right-left)/2

		if int64(potions[mid]*spell) < success {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	if int64(potions[mid]*spell) > success {
		return len(potions) - mid
	} else if int64(potions[mid]*spell) < success {
		return len(potions) - mid - 1
	}

	for mid > 0 && potions[mid-1] == potions[mid] {
		mid--
	}

	return len(potions) - mid
}

func main() {
	utils.RunTests([]utils.TestCase[[]int]{
		{
			Input:    []any{[]int{5, 1, 3}, []int{1, 2, 3, 4, 5}, int64(7)},
			Got:      successfulPairs([]int{5, 1, 3}, []int{1, 2, 3, 4, 5}, 7),
			Expected: []int{4, 0, 3},
		},
		{
			Input:    []any{[]int{5, 1, 3}, []int{1, 2, 3, 3, 4, 5}, int64(7)},
			Got:      successfulPairs([]int{5, 1, 3}, []int{1, 2, 3, 3, 4, 5}, 7),
			Expected: []int{5, 0, 4},
		},
		{
			Input:    []any{[]int{3, 1, 2}, []int{8, 5, 8}, int64(16)},
			Got:      successfulPairs([]int{3, 1, 2}, []int{8, 5, 8}, 16),
			Expected: []int{2, 0, 2},
		},
		{
			Input: []any{
				[]int{40, 11, 24, 28, 40, 22, 26, 38, 28, 10, 31, 16, 10, 37, 13, 21, 9, 22, 21, 18, 34, 2, 40, 40, 6, 16, 9, 14, 14, 15, 37, 15, 32, 4, 27, 20, 24, 12, 26, 39, 32, 39, 20, 19, 22, 33, 2, 22, 9, 18, 12, 5},
				[]int{31, 40, 29, 19, 27, 16, 25, 8, 33, 25, 36, 21, 7, 27, 40, 24, 18, 26, 32, 25, 22, 21, 38, 22, 37, 34, 15, 36, 21, 22, 37, 14, 31, 20, 36, 27, 28, 32, 21, 26, 33, 37, 27, 39, 19, 36, 20, 23, 25, 39, 40},
				int64(600),
			},
			Got: successfulPairs(
				[]int{40, 11, 24, 28, 40, 22, 26, 38, 28, 10, 31, 16, 10, 37, 13, 21, 9, 22, 21, 18, 34, 2, 40, 40, 6, 16, 9, 14, 14, 15, 37, 15, 32, 4, 27, 20, 24, 12, 26, 39, 32, 39, 20, 19, 22, 33, 2, 22, 9, 18, 12, 5},
				[]int{31, 40, 29, 19, 27, 16, 25, 8, 33, 25, 36, 21, 7, 27, 40, 24, 18, 26, 32, 25, 22, 21, 38, 22, 37, 34, 15, 36, 21, 22, 37, 14, 31, 20, 36, 27, 28, 32, 21, 26, 33, 37, 27, 39, 19, 36, 20, 23, 25, 39, 40},
				600,
			),
			Expected: []int{48, 0, 32, 37, 48, 22, 33, 47, 37, 0, 43, 6, 0, 46, 0, 21, 0, 22, 21, 14, 46, 0, 48, 48, 0, 6, 0, 0, 0, 3, 46, 3, 45, 0, 34, 20, 32, 0, 33, 47, 45, 47, 20, 18, 22, 45, 0, 22, 0, 14, 0, 0},
		},
	})
}

File: 2300/main.ts
import { runTests } from "../utils/utils.ts"

function successfulPairs(spells: number[], potions: number[], success: number): number[] {
    potions.sort((a, b) => a - b)
    let result = new Array<number>(spells.length).fill(0)

    for (let i = 0; i < spells.length; i++) {
        let start = 0
        let end = potions.length - 1
        let mid = 0
        while (start <= end) {
            mid = start + Math.floor((end - start) / 2)
            if (spells[i] * potions[mid] < success) {
                start = mid + 1
            } else {
                end = mid - 1
            }
        }
        if (potions[mid] * spells[i] >= success) {
            result[i] = potions.length - mid
        } else if (potions[mid] * spells[i] < success) {
            result[i] = potions.length - mid - 1
        }
    }

    return result
}

runTests(successfulPairs, [
    { input: [[5, 1, 3], [1, 2, 3, 4, 5], 7], expected: [4, 0, 3] },
    { input: [[3, 1, 2], [8, 5, 8], 16], expected: [2, 0, 2] },
    { input: [[1, 2, 3, 4, 5, 6, 7], [1, 2, 3, 4, 5, 6, 7], 25], expected: [0, 0, 0, 1, 3, 3, 4] },
    {
        input: [
            [15, 39, 38, 35, 33, 25, 31, 12, 40, 27, 29, 16, 22, 24, 7, 36, 29, 34, 24, 9, 11, 35, 21, 3, 33, 10, 9, 27, 35, 17, 14, 3, 35, 35, 39, 23, 35, 14, 31, 7],
            [25, 19, 30, 37, 14, 30, 38, 22, 38, 38, 26, 33, 34, 23, 40, 28, 15, 29, 36, 39, 39, 37, 32, 38, 8, 17, 39, 20, 4, 39, 39, 7, 30, 35, 29, 23],
            317
        ],
        expected: [28, 33, 33, 33, 33, 33, 33, 23, 34, 33, 33, 29, 32, 33, 0, 33, 33, 33, 33, 13, 22, 33, 31, 0, 33, 17, 13, 33, 33, 30, 27, 0, 33, 33, 33, 33, 33, 27, 33, 0]
    },
])
